internal/experimental/registry/client.go

```diff
@@ -17,325 +17,476 @@ limitations under the License.
+	return func(client *Client) {
+		client.credentialsFile = credentialsFile
 	}
-	fmt.Fprintln(c.out, "Logout succeeded")
-	return nil
 }
 
-// PushChart uploads a chart to a registry
-func (c *Client) PushChart(ref *Reference) error {
-	r, err := c.cache.FetchReference(ref)
-	if err != nil {
-		return err
+type (
+	// LoginOption allows specifying various settings on login
+	LoginOption func(*loginOperation)
+
+	// LoginResult is the result returned upon successful login
+	LoginResult struct {
+		Host string `json:"host"`
 	}
-	if !r.Exists {
-		return errors.New(fmt.Sprintf("Chart not found: %s", r.Name))
+
+	loginOperation struct {
+		username string
+		password string
+		insecure bool
 	}
-	fmt.Fprintf(c.out, "The push refers to repository [%s]\n", r.Repo)
-	c.printCacheRefSummary(r)
-	layers := []ocispec.Descriptor{*r.ContentLayer}
-	_, err = oras.Push(ctx(c.out, c.debug), c.resolver, r.Name, c.cache.Provider(), layers,
-		oras.WithConfig(*r.Config), oras.WithNameValidation(nil))
+)
+
+// Login logs into a registry
+func (c *Client) Login(host string, options ...LoginOption) (*LoginResult, error) {
+	operation := &loginOperation{}
+	for _, option := range options {
+		option(operation)
+	}
+	authorizerLoginOpts := []auth.LoginOption{
+		auth.WithLoginContext(ctx(c.out, c.debug)),
+		auth.WithLoginHostname(host),
+		auth.WithLoginUsername(operation.username),
+		auth.WithLoginSecret(operation.password),
+		auth.WithLoginUserAgent(version.GetUserAgent()),
+	}
+	if operation.insecure {
+		authorizerLoginOpts = append(authorizerLoginOpts, auth.WithLoginInsecure())
+	}
+	err := c.authorizer.LoginWithOpts(authorizerLoginOpts...)
 	if err != nil {
-		return err
+		return nil, err
+	}
+	result := &LoginResult{
+		Host: host,
 	}
-	s := ""
-	numLayers := len(layers)
-	if 1 < numLayers {
-		s = "s"
+	fmt.Fprintln(c.out, "Login Succeeded")
+	return result, nil
+}
+
+// LoginOptBasicAuth returns a function that sets the username/password settings on login
+func LoginOptBasicAuth(username string, password string) LoginOption {
+	return func(operation *loginOperation) {
+		operation.username = username
+		operation.password = password
 	}
-	fmt.Fprintf(c.out,
-		"%s: pushed to remote (%d layer%s, %s total)\n", r.Tag, numLayers, s, byteCountBinary(r.Size))
-	return nil
 }
 
-// PullChart downloads a chart from a registry
-func (c *Client) PullChart(ref *Reference) (*bytes.Buffer, error) {
-	buf := bytes.NewBuffer(nil)
+// LoginOptInsecure returns a function that sets the insecure setting on login
+func LoginOptInsecure(insecure bool) LoginOption {
+	return func(operation *loginOperation) {
+		operation.insecure = insecure
+	}
+}
 
-	if ref.Tag == "" {
-		return buf, errors.New("tag explicitly required")
+type (
+	// LogoutOption allows specifying various settings on logout
+	LogoutOption func(*logoutOperation)
+
+	// LogoutResult is the result returned upon successful logout.
+	LogoutResult struct {
```

Comment:
What are the LogoutResult and LoginResult types used for?

Reply:
While they are not useful today, consider a future registry login which returns some sort of token etc. back to the user.

In general, I tried to follow the same pattern for all registry actions (`Push`, `Pull`, `Login`, `Logout`), where they each return a custom "Result" type. I'm attempting to future-proof this design with options and results.

I'm anticipating preventing the need to add methods such as `LoginWithToken` and deprecating `Login` when/if that time comes (which I've seen done a number of times in this codebase).

Reply:
Similarly, you'll notice  the signature of `Logout` is:
```go
func (c *Client) Logout(host string, opts ...LogoutOption) (*LogoutResult, error)
```

even though there are no such `LogoutOption`'s defined yet.

Never want to have to break this API!

Reply:
If the idea is to return a "result" based on a generic login/logout implementation, why not return an interface instead? That way you don't have a never-ending struct with multiple fields handling different protocols like `Username/Password` and `AuthToken`. Instead you could return a `Result` type with function definitions like `Identity() string` that returns relevant information to an identity provider.

https://golang.org/doc/effective_go#interfaces_and_types

Reply:
Resolved in 8202ccd6998e6bf8aa34c8d9d7527e08d326397b